examples/chat/src/util/auth.rs

```diff
@@ -0,0 +1,67 @@
+use http::Extensions;
+use via::request::RequestHead;
+use via::{Middleware, raise, ws};
+
+use crate::chat::Chat;
+use crate::models::user::User;
+use crate::{Next, Request};
+
+pub trait Authenticate {
+    fn current_user(&self) -> via::Result<&User>;
+}
+
+pub struct Auth {
+    cookie: String,
+}
+
+#[derive(Clone)]
+struct Verify(User);
+
+#[inline]
+fn try_from_extensions(extensions: &Extensions) -> via::Result<&User> {
+    let Some(Verify(user)) = extensions.get() else {
+        raise!(401);
+    };
+
+    Ok(user)
+}
+
+impl Auth {
+    pub fn new(cookie: impl Into<String>) -> Self {
+        Self {
+            cookie: cookie.into(),
+        }
+    }
+}
+
+impl Middleware<Chat> for Auth {
+    fn call(&self, mut request: Request, next: Next) -> via::BoxFuture {
+        let jar = request.cookies().private(request.state().secret());
+
+        if let Some(cookie) = jar.get(&self.cookie)
+            && let Ok(user) = serde_json::from_str(cookie.value())
+        {
+            request.extensions_mut().insert(Verify(user));
+        }
+
+        next.call(request)
+    }
+}
+
+impl Authenticate for Request {
```

Comment:
I'm considering making request a container type. Users would have to either take a reference to the request head or move the request head and body out of request before accessing the fields they want to access.

It sounds cumbersome but the explicit moves in the route modules are really nice to read and work with.